contracts/auto-withdrawer/src/contract.rs

```diff
@@ -226,33 +266,123 @@ pub fn reply(
+        .flat_map(|event| event.attributes.iter())
+        .find(|attribute| attribute.key == key)
+        .unwrap()
+        .value
+        .clone()
+}
+
+fn merge_deposits(vec1: Vec<Coin>, vec2: Vec<Coin>) -> Vec<Coin> {
+    let mut coin_map: HashMap<String, Uint128> = HashMap::new();
+
+    for coin in vec1.into_iter().chain(vec2.into_iter()) {
+        coin_map
+            .entry(coin.denom)
+            .and_modify(|e| *e += coin.amount)
+            .or_insert(coin.amount);
+    }
+
+    let mut merged_coins: Vec<Coin> = coin_map
+        .into_iter()
+        .map(|(denom, amount)| Coin { denom, amount })
+        .collect();
+
+    merged_coins.sort_by(|a, b| a.denom.cmp(&b.denom));
+
+    merged_coins
+}
+
+fn divide_deposits(coins: Vec<Coin>, multiplier: Decimal) -> Vec<Coin> {
+    coins
+        .into_iter()
+        .map(|coin| Coin {
+            denom: coin.denom,
+            amount: coin.amount * multiplier,
+        })
+        .collect()
+}
+
+fn subtract_deposits(vec1: Vec<Coin>, vec2: Vec<Coin>) -> Vec<Coin> {
+    let mut coin_map: HashMap<String, Uint128> = HashMap::new();
+
+    for coin in vec1 {
+        coin_map.insert(coin.denom.clone(), coin.amount);
+    }
+
+    for coin in vec2 {
+        coin_map
+            .entry(coin.denom.clone())
+            .and_modify(|e| {
+                if *e >= coin.amount {
+                    *e -= coin.amount;
+                }
+            })
+            .or_insert(Uint128::zero());
     }
+
+    let mut result_coins: Vec<Coin> = coin_map
+        .into_iter()
+        .filter(|(_, amount)| !amount.is_zero())
+        .map(|(denom, amount)| Coin { denom, amount })
+        .collect();
+
+    result_coins.sort_by(|a, b| a.denom.cmp(&b.denom));
+
+    result_coins
+}
+
+fn get_bonding_id(sender: impl Display, batch_id: impl Display) -> String {
+    format!("{sender}_{batch_id}")
+}
+
+fn get_full_withdrawal_denom(
+    withdrawal_denom_prefix: impl Display,
+    withdrawal_token_address: impl Display,
+    batch_id: Uint128,
+) -> String {
+    format!("factory/{withdrawal_token_address}/{withdrawal_denom_prefix}:{UNBOND_MARK}:{batch_id}")
 }
 
 fn reply_core_unbond(
     deps: DepsMut<NeutronQuery>,
     sender: Addr,
-    deposit: Vec<Coin>,
+    mut deposit: Vec<Coin>,
     events: Vec<Event>,
 ) -> ContractResult<Response<NeutronMsg>> {
-    let token_id = events
-        .into_iter()
-        .filter(|event| event.ty == "wasm")
-        .flat_map(|event| event.attributes)
-        .find(|attribute| attribute.key == "token_id")
-        // it is safe to use unwrap here because cw-721 always generates valid events on success
-        .unwrap()
-        .value;
+    let batch_id = get_core_event_value(&events, "batch_id");
+    let str_amount = get_core_event_value(&events, "amount");
+
+    let mut amount = match Uint128::from_str(&str_amount) {
+        Ok(value) => Ok(value),
+        Err(_) => Err(ContractError::InvalidCoreReplyAttributes {}),
```

Comment:
Same here. Since we serialize this data ourselves with our own serializer, we can safely deserialize.unwrap() it here, without spawning more errors than needed.

Reply:
it is a part of a reply handler which theoretically can be called with incorrect data

Reply:
The same answer as above, the data passed is provided by core contract, which is controlled by us. If Neutron binary suddenly decides to mess up with our event logs, then we wouldn't be able to do **anything** about.